Container resolvers: add docs, typing and tests (#15614)
galaxyproject/galaxy: doc/source/admin/jobs.md

the user that Galaxy runs as should be able to run ``sudo docker`` without a password prompt for Docker containers to
work.

The images used for containers can either be specified explicitly in the ``<destination>`` using the ``docker_default_container_id``, ``docker_container_id_override``, ``singularity_default_container_id`` and
``singularity_container_id_override`` parameters, but (perhaps more commonly) the image to use can be derived from the
``<requirements>`` of the Galaxy tool being executed. In the latter case the image is specified either explicitly
using a ``<container>`` tag or a mulled (multi package) container is implied for the set 
of packages specified by the tool's ``<requirement>`` tags. 
In either case the container to be used is determined using container resolvers that can be specified

> globally for an instance of locally per instance, see [Container resolvers](container_resolvers).

Reviewer: ```suggestion
globally for an instance and/or per execution environment, see [Container resolvers](container_resolvers).
```

> globally for an instance of locally per instance, see [Container resolvers](container_resolvers.md).

Reviewer: ```suggestion
globally for an instance of locally per instance, see [Container resolvers](container_resolvers).
```

> globally for an instance of locally per instrance, see [Container resolvers](container_resolvers.md).

Reviewer: This line is unclear.

> globally for an instance of locally per instrance, see [Container resolvers](container_resolvers.md).

Author: I think I wanted to write:

```suggestion
globally for an instance and/or per execution environment, see [Container resolvers](container_resolvers.md).
```

Does this make more sense?

I also realized that I often wrote compute environment instead of execution environment .. I will fix this.

I just have to note once that I don't really like that destinations have been renamed to execution environments. IMO the not (really clearly) speaking name was not the real problem, but the lack of definition what a destination/execution environment is. This is still missing (but I have to admit that the name is a bit better) and we have inconsistent naming in docs and code :( .. maybe it's my formal/math background that suggests me that a name is not important as long as it is defined and used consistently.

Reviewer: > Does this make more sense?

Yes, thanks, can you please make the change?
